nerfstudio/generative/stable_diffusion.py

```diff
@@ -0,0 +1,343 @@
+# Copyright 2022 The Nerfstudio Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Stable Diffusion helpers"""
+
+# Modified from https://github.com/ashawkey/stable-dreamfusion/blob/main/nerf/sd.py
+
+import sys
+from pathlib import Path
+from typing import List, Optional, Tuple, Union
+
+import mediapy
+import numpy as np
+import torch
+import torch.nn.functional as F
+import tyro
+from rich.console import Console
+from torch import nn
+from torchtyping import TensorType
+
+CONSOLE = Console(width=120)
+
+try:
+    from diffusers import PNDMScheduler, StableDiffusionPipeline
+    from transformers import logging
+
+except ImportError:
+    CONSOLE.print("[bold red]Missing Stable Diffusion packages.")
+    CONSOLE.print(r"Install using [yellow]pip install nerfstudio\[gen][/yellow]")
+    CONSOLE.print(r"or [yellow]pip install -e .\[gen][/yellow] if installing from source.")
+    sys.exit(1)
+
+logging.set_verbosity_error()
+IMG_DIM = 512
+CONST_SCALE = 0.18215
+
+SD_SOURCE = "runwayml/stable-diffusion-v1-5"
+CLIP_SOURCE = "openai/clip-vit-large-patch14"
+
+
+class StableDiffusion(nn.Module):
+    """Stable Diffusion implementation
+
+    Args:
+        device: device to use
+        num_train_timesteps: number of training timesteps
+
+    """
+
+    def __init__(self, device: Union[torch.device, str], num_train_timesteps: int = 1000) -> None:
+        super().__init__()
+
+        self.device = device
+        self.num_train_timesteps = num_train_timesteps
+
+        self.min_step = int(self.num_train_timesteps * 0.02)
+        self.max_step = int(self.num_train_timesteps * 0.98)
+
+        self.scheduler = PNDMScheduler(
+            beta_start=0.00085,
+            beta_end=0.012,
+            beta_schedule="scaled_linear",
+            num_train_timesteps=self.num_train_timesteps,
+        )
+        self.alphas = self.scheduler.alphas_cumprod.to(self.device)  # type: ignore
+
+        pipe = StableDiffusionPipeline.from_pretrained(SD_SOURCE, torch_dtype=torch.float16)
+        assert pipe is not None
+        pipe = pipe.to(self.device)
+
+        # MEMORY IMPROVEMENTS
+        pipe.enable_attention_slicing()
```

Comment:
This is the only memory optimization we ended up needing, should we get rid of the rest?